### ISSUE TITLE
Missing ServiceAccount, some Role rules and RoleBinding with OLM installation

### ISSUE THREAD

**Issue by rascasoft** [aquasecurity/trivy-operator#1584]
# Description

Hi all,
I'm Following the procedure described in [the official docs](https://aquasecurity.github.io/trivy-operator/v0.1.4/operator/installation/olm/), and after creating the `Subscription` resource the status of the cluster service version CR is `Pending`:

```console
$ kubectl get clusterserviceversions.operators.coreos.com -n trivy-system
NAME                     DISPLAY          VERSION   REPLACES                 PHASE
trivy-operator.v0.16.3   Trivy Operator   0.16.3    trivy-operator.v0.16.2   Pending
```

The initial reason is the absence of the `ServiceAccount` resource named `trivy-operator`:

```console
$ kubectl describe clusterserviceversions.operators.coreos.com -n trivy-system
...
...
    Group:    
    Kind:     ServiceAccount
    Message:  Service account does not exist
    Name:     trivy-operator
    Status:   NotPresent
    Version:  v1
Events:
  Type    Reason               Age   From                        Message
  ----    ------               ----  ----                        -------
  Normal  RequirementsUnknown  10m   operator-lifecycle-manager  requirements not yet checked
  Normal  RequirementsNotMet   10m   operator-lifecycle-manager  one or more requirements couldn't be found
```

But even after creating it, some permissions are missing in the `ClusterRole` named `trivy-operator`:

```console
$ kubectl describe clusterserviceversions.operators.coreos.com -n trivy-system
...
...
      Group:    rbac.authorization.k8s.io
      Kind:     PolicyRule
      Message:  cluster rule:{"verbs":["get","list","watch","create","update","delete"],"apiGroups":["aquasecurity.github.io"],"resources":["vulnerabilityreports","configauditreports","clustercompliancereports","clusterconfigauditreports","exposedsecretreports","sbomreports","rbacassessmentreports","infraassessmentreports","clusterrbacassessmentreports","clusterinfraassessmentreports"]}
      Status:   NotSatisfied
      Version:  v1
      Group:    rbac.authorization.k8s.io
      Kind:     PolicyRule
      Message:  cluster rule:{"verbs":["create","get","update"],"apiGroups":["coordination.k8s.io"],"resources":["leases"]}
      Status:   NotSatisfied
      Version:  v1
    Group:      
    Kind:       ServiceAccount
    Message:    Policy rule not satisfied for service account
    Name:       trivy-operator
    Status:     PresentNotSatisfied
    Version:    v1
Events:
  Type    Reason               Age   From                        Message
  ----    ------               ----  ----                        -------
  Normal  RequirementsUnknown  0s    operator-lifecycle-manager  requirements not yet checked
  Normal  RequirementsNotMet   0s    operator-lifecycle-manager  one or more requirements couldn't be found
```

After manually adding everything, the Operator gets installed, so I'm using this yaml that contains what is missing:

```yaml
# ServiceAccount
apiVersion: v1
kind: ServiceAccount
metadata:
  name: trivy-operator
  namespace: trivy-system
  labels:
    app.kubernetes.io/name: trivy-operator
    app.kubernetes.io/instance: trivy-operator
    app.kubernetes.io/version: "0.1.4"
    app.kubernetes.io/managed-by: kubectl
---
# ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  creationTimestamp: null
  name: trivy-operator
rules:
####################################################
# These are the installed rules:
- apiGroups:
  - ""
  resources:
  - configmaps
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - limitranges
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - pods/log
  verbs:
  - get
  - list
- apiGroups:
  - ""
  resources:
  - replicationcontrollers
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - resourcequotas
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - secrets
  verbs:
  - create
  - get
- apiGroups:
  - ""
  resources:
  - serviceaccounts
  verbs:
  - get
- apiGroups:
  - ""
  resources:
  - services
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - apiextensions.k8s.io
  resources:
  - customresourcedefinitions
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - apps
  resources:
  - daemonsets
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - apps
  resources:
  - replicasets
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - apps
  resources:
  - statefulsets
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - aquasecurity.github.io
  resources:
  - clustercompliancedetailreports
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - aquasecurity.github.io
  resources:
  - clustercompliancereports
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - aquasecurity.github.io
  resources:
  - clustercompliancereports/status
  verbs:
  - get
  - patch
  - update
- apiGroups:
  - aquasecurity.github.io
  resources:
  - clusterconfigauditreports
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - aquasecurity.github.io
  resources:
  - clusterrbacassessmentreports
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - aquasecurity.github.io
  resources:
  - configauditreports
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - aquasecurity.github.io
  resources:
  - exposedsecretreports
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - aquasecurity.github.io
  resources:
  - rbacassessmentreports
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - aquasecurity.github.io
  resources:
  - vulnerabilityreports
  verbs:
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - batch
  resources:
  - cronjobs
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - batch
  resources:
  - jobs
  verbs:
  - create
  - delete
  - get
  - list
  - watch
- apiGroups:
  - networking.k8s.io
  resources:
  - ingresses
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - networking.k8s.io
  resources:
  - networkpolicies
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - rbac.authorization.k8s.io
  resources:
  - clusterrolebindings
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - rbac.authorization.k8s.io
  resources:
  - clusterroles
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - rbac.authorization.k8s.io
  resources:
  - rolebindings
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - rbac.authorization.k8s.io
  resources:
  - roles
  verbs:
  - get
  - list
  - watch
####################################################
# These are the missing rules that I manually added:
- verbs:
    - get
    - list
    - watch
  apiGroups:
    - ""
  resources:
    - pods
    - pods/log
    - replicationcontrollers
    - services
    - resourcequotas
    - limitranges
- verbs:
    - get
    - list
    - watch
  apiGroups:
    - ""
  resources:
    - nodes
- verbs:
    - list
    - watch
    - get
    - create
    - update
  apiGroups:
    - ""
  resources:
    - configmaps
    - secrets
    - serviceaccounts
- verbs:
    - delete
  apiGroups:
    - ""
  resources:
    - secrets
- verbs:
    - create
  apiGroups:
    - ""
  resources:
    - events
- verbs:
    - get
    - list
    - watch
  apiGroups:
    - apps
  resources:
    - replicasets
    - statefulsets
    - daemonsets
    - deployments
- verbs:
    - get
    - list
    - watch
  apiGroups:
    - policy
  resources:
    - podsecuritypolicies
- verbs:
    - get
    - list
    - watch
    - create
    - update
    - delete
  apiGroups:
    - aquasecurity.github.io
  resources:
    - vulnerabilityreports
    - configauditreports
    - clustercompliancereports
    - clusterconfigauditreports
    - exposedsecretreports
    - sbomreports
    - rbacassessmentreports
    - infraassessmentreports
    - clusterrbacassessmentreports
    - clusterinfraassessmentreports
- verbs:
    - create
    - get
    - update
  apiGroups:
    - coordination.k8s.io
  resources:
    - leases
 ---
# ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: trivy-operator
  labels:
    app.kubernetes.io/name: trivy-operator
    app.kubernetes.io/instance: trivy-operator
    app.kubernetes.io/version: "0.1.4"
    app.kubernetes.io/managed-by: kubectl
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: trivy-operator
subjects:
  - kind: ServiceAccount
    name: trivy-operator
    namespace: trivy-system
```

## Issue

I was expecting everything to be covered by the subscription, but this also makes me feel like I'm missing something. Is this a normal behavior? Is it expected that the admin must create the SA/Role/RoleBinding by himself before creating the Subscription?

This doesn't happen while installing everything using just `kubectl`.

## Environment

- Trivy-Operator version (use `trivy-operator version`): 0.16.3
- Kubernetes version (use `kubectl version`): v1.27.4
- OS (macOS 10.15, Windows 10, Ubuntu 19.10 etc): AlmaLinux release 8.8 (Sapphire Caracal)

**Comment by chen-keinan** [comment]
@rascasoft has is a [ref for operator](https://github.com/k8s-operatorhub/community-operators/blob/main/operators/trivy-operator/0.16.4/manifests/trivy-operator.v0.16.4.clusterserviceversion.yaml#L76) I'll take a look at it as well.

also please review [operator install instruction](https://artifacthub.io/packages/olm/community-operators/trivy-operator?modal=install)

**Comment by rascasoft** [comment]
Thanks for your answer @chen-keinan , but I didn't understand what you're trying to tell me.
The first link is the manifest of the cluster service version of the trivy-operator in the operatorhub sources, and I'm not able to understand from here what I want to know, which is if it's on purpose that service account and role/rolebindings are not created by the operator installer.
The second link takes me to the instructions which I already followed and don't tell me much more compared to what is the doc I followed (this one doesn't cover the operatorgroup resource creation for instance).

Am I missing something?

Many thanks!

**Comment by chen-keinan** [comment]
> Thanks for your answer @chen-keinan , but I didn't understand what you're trying to tell me. The first link is the manifest of the cluster service version of the trivy-operator in the operatorhub sources, and I'm not able to understand from here what I want to know, which is if it's on purpose that service account and role/rolebindings are not created by the operator installer. The second link takes me to the instructions which I already followed and don't tell me much more compared to what is the doc I followed (this one doesn't cover the operatorgroup resource creation for instance).
> 
> Am I missing something?
> 
> Many thanks!

its just a ref for operator hub OLM official install instructions, I'll have another look myself and see if I can find the issue

**Comment by rascasoft** [comment]
Many thanks @chen-keinan. There's one main thing I would like to understand (and maybe you can help here): when I configure a subscription for OLM, are all the resources that are going to be used (so Service Accounts, ClusterRole, ClusterRoleBinding, etc.) expected to be created by the subscription itself?

I ask this because if you use OLM in OpenShift, then everything works flawlessly, but in K8s things seem different.

Many thanks!

**Comment by chen-keinan** [comment]
> Many thanks @chen-keinan. There's one main thing I would like to understand (and maybe you can help here): when I configure a subscription for OLM, are all the resources that are going to be used (so Service Accounts, ClusterRole, ClusterRoleBinding, etc.) expected to be created by the subscription itself?
> 
> I ask this because if you use OLM in OpenShift, then everything works flawlessly, but in K8s things seem different.
> 
> Many thanks!

yes , this is how it works based on my understanding

**Comment by rascasoft** [comment]
I've got an addition to made at this, after further testing: it seems that if you install OLM using `operator-sdk` the behavior changes to what you expect.
On a clean environment I installed OLM by using:

`$ operator-sdk olm install`

And then I created `OperaotrGroup` and `Subscription` as follows:

```yaml
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: trivy-operator-group
  namespace: trivy-system
spec:
  targetNamespaces:
    - trivy-system
    - mions1
---
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: trivy-operator-subscription
  namespace: trivy-system
spec:
  channel: alpha
  name: trivy-operator
  source: operatorhubio-catalog
  sourceNamespace: olm
  installPlanApproval: Automatic
  config:
    env:
    - name: OPERATOR_EXCLUDE_NAMESPACES
      value: "kube-system,trivy-system
```

The CSV installed correctly:

```console
$ kubectl get csv -n trivy-system 
NAME                     DISPLAY          VERSION   REPLACES                 PHASE
trivy-operator.v0.16.4   Trivy Operator   0.16.4    trivy-operator.v0.16.3   Succeeded
```

The only remaining problem was the absence of the `trivy-operator-trivy-config` config map inside the `trivy-system` namespace, which was causing this message to appear in the pod logs:

```console
{"level":"error","ts":"2023-11-14T16:53:09Z","msg":"Reconciler error","controller":"clusterrole","controllerGroup":"rbac.authorization.k8s.io","controllerKind":"ClusterRole","ClusterRole":{"name":"rbacassessmentreports.aquasecurity.github.io-v1alpha1-crdview"},"namespace":"","name":"rbacassessmentreports.aquasecurity.github.io-v1alpha1-crdview","reconcileID":"4f4e747e-6947-416c-bada-22e270b1f9fe","error":"ConfigMap \"trivy-operator-trivy-config\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227"}
```

So I created it manually, using [the code from the static manifest](https://github.com/aquasecurity/trivy-operator/blob/54e6c909263e0394cb67dedc2c44bbfc06216894/deploy/static/trivy-operator.yaml#L2471-L2504):

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: trivy-operator-trivy-config
  namespace: trivy-system
  labels:
    app.kubernetes.io/name: trivy-operator
    app.kubernetes.io/instance: trivy-operator
    app.kubernetes.io/version: "0.16.4"
    app.kubernetes.io/managed-by: kubectl
data:
  trivy.repository: "ghcr.io/aquasecurity/trivy"
  trivy.tag: "0.47.0"
  trivy.imagePullPolicy: "IfNotPresent"
  trivy.additionalVulnerabilityReportFields: ""
  trivy.severity: "UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL"
  trivy.slow: "true"
  trivy.skipJavaDBUpdate: "false"
  trivy.imageScanCacheDir: "/tmp/trivy/.cache"
  trivy.filesystemScanCacheDir: "/var/trivyoperator/trivy-db"
  trivy.dbRepository: "ghcr.io/aquasecurity/trivy-db"
  trivy.javaDbRepository: "ghcr.io/aquasecurity/trivy-java-db"
  trivy.command: "image"
  trivy.dbRepositoryInsecure: "false"
  trivy.useBuiltinRegoPolicies: "true"
  trivy.supportedConfigAuditKinds: "Workload,Service,Role,ClusterRole,NetworkPolicy,Ingress,LimitRange,ResourceQuota"                                                    
  trivy.timeout: "5m0s"
  trivy.mode: "Standalone"
  trivy.resources.requests.cpu: "100m"
  trivy.resources.requests.memory: "100M"
  trivy.resources.limits.cpu: "500m"
  trivy.resources.limits.memory: "500M"
```

Even if I'm still getting some weird logs inside the Pod:

```console
{"level":"error","ts":"2023-11-14T16:55:40Z","msg":"Reconciler error","controller":"job","controllerGroup":"batch","controllerKind":"Job","Job":{"name":"scan-vulnerabilityreport-75dc95d75c","namespace":"trivy-system"},"namespace":"trivy-system","name":"scan-vulnerabilityreport-75dc95d75c","reconcileID":"94e371e7-4f8e-4499-b918-39945792221f","error":"SbomReport.aquasecurity.github.io \"replicaset-nginx-8bc664746-nginx\" is invalid: report.components.specVersion: Invalid value: \"string\": report.components.specVersion in body must be of type integer: \"string\"","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227"}                      
```

The `configauditreport` and `vulnerabilityreport` are generated.

So I see two topics here:

1. Understand what is the difference between manual and `operator-sdk` OLM installation (not to be addressed here).
2. Fix the ConfigMap generation inside the trivy-operator manifest (which becomes the main issue of this bug).

**Comment by chen-keinan** [comment]
Thanks, i'll have a look at this feedback

**Comment by chen-keinan** [comment]
@chen-keinan I have created a patch , it will be released with `trivy-operator v0.17.0`
